File: Java/testMatrixMultiplication.java
import java.util.Scanner;
import java.io.*;
import java.util.concurrent.*;
import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.Random;

class testMatrixMultiplication {

    public static void main (String[] args) {

        long startTime;
        long estimatedTime;
        int n;

        // create a scanner so we can read the command-line input
        Scanner scanner = new Scanner(System.in);

        // create instance of Random class
        Random rand = new Random();

        if (args.length > 0) {
            try {
                n = Integer.parseInt(args[0]);
            } catch (NumberFormatException e) {
                System.err.println("Argument" + args[0] + " must be an integer.");
                System.exit(1);
            }
        }

        n = Integer.parseInt(args[0]);
        double a[][]=new double[n][n];
        double b[][]=new double[n][n];
        double c[][]=new double[n][n];

        for (int i=0;i<n;i++) {
            for (int j=0;j<n;j++) {
                a[i][j] = rand.nextDouble();
                b[i][j] = rand.nextDouble();
            }
        }

        System.out.println("Matrix Multiplication: "+n);

        startTime = System.currentTimeMillis();
        c = multiply(a, b);
        estimatedTime = System.currentTimeMillis() - startTime;
        NumberFormat formatter = new DecimalFormat("#00000.00000000");
        System.out.println("    Time for matMult: " + formatter.format(estimatedTime  / 1000d) + " seconds");

    }

    // return c = a * b
    public static double[][] multiply(double[][] a, double[][] b) {
        int m1 = a.length;
        int n1 = a[0].length;
        int m2 = b.length;
        int n2 = b[0].length;
        if (n1 != m2) throw new RuntimeException("Illegal matrix dimensions.");
        double[][] c = new double[m1][n2];
        for (int i = 0; i < m1; i++)
            for (int j = 0; j < n2; j++)
                for (int k = 0; k < n1; k++)
                    c[i][j] += a[i][k] * b[k][j];
        return c;
    }
}

File: Rust/test_matrix_multiplication.rs

use rand::Rng;
use std::env;
use std::process;
use std::time::Instant;

fn main() {
    // Collect command line arguments
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <dim>", args[0]);
        process::exit(1);
    }

    // Parse the dimension from the command line
    let dim: usize = match args[1].parse() {
        Ok(num) => num,
        Err(_) => {
            eprintln!("Error: dim must be a valid integer.");
            process::exit(1);
        }
    };

    // Allocate the 2D arrays
    let mut a = make_2d_array_double(dim, dim);
    let mut b = make_2d_array_double(dim, dim);
    let mut c = make_2d_array_double(dim, dim);

    let mut rng = rand::rng();

    // Fill matrices A and B with random numbers between 0.0 and 1.0
    for i in 0..dim {
        for j in 0..dim {
            a[i][j] = rng.random::<f64>();
            b[i][j] = rng.random::<f64>();
        }
    }

    // Start the timer
    let start = Instant::now();

    // Perform the matrix multiplication (C = A * B)
    for i in 0..dim {
        for j in 0..dim {
            c[i][j] = 0.0;
        }
        for k in 0..dim {
            for j in 0..dim {
                c[i][j] += a[i][k] * b[k][j];
            }
        }
    }

    // End the timer
    let duration = start.elapsed();

    println!(
        "time for C({dim},{dim}) = A({dim},{dim}) B({dim},{dim}) is {:.6} s",
        duration.as_secs_f64()
    );
}

/// Helper function to create a 2D vector initialized with zeros.
/// This replaces the nested mallocs from the C `make_2d_array_double` function.
fn make_2d_array_double(array_size_x: usize, array_size_y: usize) -> Vec<Vec<f64>> {
    vec![vec![0.0f64; array_size_y]; array_size_x]
}
